fluentinterface/src/main/java/com/iluwatar/fluentinterface/fluentiterable/lazy/LazyFluentIterable.java

```diff
@@ -0,0 +1,236 @@
+import java.util.Optional;
+import java.util.function.Consumer;
+import java.util.function.Function;
+import java.util.function.Predicate;
+
+/**
+ * This is a lazy implementation of the FluentIterable interface. It evaluates
+ * all chained operations when a terminating operation is applied.
+ * @param <TYPE> the type of the objects the iteration is about
+ */
+public class LazyFluentIterable<TYPE> implements FluentIterable<TYPE> {
+
+    private final Iterable<TYPE> iterable;
+
+    /**
+     * This constructor creates a new LazyFluentIterable. It wraps the
+     * given iterable.
+     * @param iterable the iterable this FluentIterable works on.
+     */
+    protected LazyFluentIterable(Iterable<TYPE> iterable) {
+        this.iterable = iterable;
+    }
+
+    /**
+     * This constructor can be used to implement anonymous subclasses
+     * of the LazyFluentIterable.
+     */
+    protected LazyFluentIterable() {
+        iterable = this;
+    }
+
+    /**
+     * Adds a filter operation to the operation chain and returns a new iterable.
+     * @param predicate the condition to test with for the filtering. If the test
+     *                  is negative, the tested object is removed by the iterator.
+     * @return a new FluentIterable object that decorates the source iterable
+     */
+    @Override
+    public FluentIterable<TYPE> filter(Predicate<? super TYPE> predicate) {
+        return new LazyFluentIterable<TYPE>() {
+            @Override
+            public Iterator<TYPE> iterator() {
+                return new DecoratingIterator<TYPE>(iterable.iterator()) {
+                    @Override
+                    public TYPE computeNext() {
+                        while(true) {
+                            if(fromIterator.hasNext()) {
+                                TYPE candidate = fromIterator.next();
+                                if(!predicate.test(candidate)) {
+                                    continue;
+                                }
+                                return candidate;
+                            }
+
+                            return null;
+                        }
+                    }
+                };
+            }
+        };
+    }
+
+    /**
+     * Uses the Iterable interface's forEach method to apply a given function
+     * for each object of the iterator. Is a terminating operation.
+     * @param action the action for each object
+     */
+    @Override
+    public void forEachDo(Consumer<? super TYPE> action) {
+        Iterator<TYPE> newIterator = iterable.iterator();
+        while(newIterator.hasNext()) {
+            action.accept(newIterator.next());
+        }
+    }
+
+    /**
+     * Can be used to collect objects from the iteration. Is a terminating operation.
+     * @return an option of the first object of the iteration
+     */
+    @Override
+    public Optional<TYPE> first() {
+        Optional result = Optional.empty();
+        List<TYPE> list = first(1).asList();
```

Comment:
hasNext has to be called on the Iterator, not the Iterable, but I changed it similar to your example. Thanks,